File: simulator/actions/melee_attack.py
import math

from cachetools import cached
from cachetools.keys import hashkey

from ..actions.actoid import FactoryFlags
from ..actions.attack import AttackFactory, Attack
from ..battle_map import Map
from ..misc import Conditions
import logging


logger = logging.getLogger("Encounterra")


class MeleeAttackFactory(AttackFactory):

    def __init__(self, name, combatant, to_hit, dmg_dice, dmg_bonus, dmg_type, attack_range, action_type, crit_range=1, ammo=math.inf, on_hit=None, extra_dmg=[], finesse=False):
        super().__init__(name, combatant, to_hit, dmg_dice, dmg_bonus, dmg_type, attack_range, action_type, crit_range, ammo, on_hit, extra_dmg)
        self.flags |= FactoryFlags.IS_MELEE
        if finesse:
            self.flags |= FactoryFlags.IS_FINESSE

    def get_ability_name(self):
        return "Melee Attack"

    def create(self, target):
        return MeleeAttack(target, self)

    def create_all(self, previous_action_in_dag=None):
        targets = self.get_eligible_targets()
        return [MeleeAttack(t, self) for t in targets]


class MeleeAttack(Attack):

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        swallower = self.factory.combatant.get_swallower()
        if swallower:
            if swallower is self.target:
                return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]
            return None
        if not self.factory.combatant.is_affected_by_any(Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            return battle_map.get_free_coords_in_hop_range(battle_map.get_combatant_position(self.target),
                                                           distances,
                                                           inflate_to_size=self.factory.combatant.size,
                                                           rng=self.factory.range,
                                                           combatant=self.factory.combatant)
        elif battle_map.are_in_hop_range(self.factory.combatant, self.target, self.factory.range):
            return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]

File: c++/actions/melee_attack.cpp
#include "actions/melee_attack.hpp"
#include "core/battle_map.hpp"
#include "core/combatant.hpp"

namespace enc
{
  MeleeAttackFactory::MeleeAttackFactory(const std::string &name, const std::string &abilityName, Combatant *combatant, AbilityType abilityType,
                                         int toHit, std::vector<Die> dmgDice, int dmgBonus, DamageType dmgType, int attackRange, int critRange,
                                         Uses &&ammo, std::vector<std::unique_ptr<OnHit>> onHit, std::vector<DmgDieWithType> extraDmg, bool usesDex,
                                         bool twoHanded, Die toHitBonusDie)
      : AttackFactory(name, abilityName, combatant, abilityType, toHit, dmgDice, dmgBonus, dmgType, attackRange, critRange, std::move(ammo),
                      std::move(onHit), extraDmg, usesDex, twoHanded, toHitBonusDie)
  {
    setFlag(FactoryFlags::IS_MELEE);
  }

  std::vector<Actoid *> MeleeAttackFactory::createAll(void *previousActionInDag)
  {
    auto eligibleTargets = getEligibleTargets();
    std::vector<Actoid *> result;
    result.reserve(eligibleTargets.size());
    for(const auto &target : eligibleTargets)
      {
        result.push_back(std::make_shared<MeleeAttack>(AbilityType::MELEE_ATTACK, *target, *this));
      }
    return result;
  }

  Actoid * MeleeAttackFactory::create(void *target) { return std::make_shared<MeleeAttack>(AbilityType::MELEE_ATTACK, *static_cast<Combatant *>(target), *this); }

  std::optional<CoordVector>
  MeleeAttack::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    MeleeAttackFactory &factory = dynamic_cast<MeleeAttackFactory &>(getFactory());
    BattleMap &battleMap = BattleMap::getInstance();

    auto combatant = factory._combatant;
    if(Combatant *swallower = combatant->getSwallower())
      {
        if(*swallower == _target)
          {
            return CoordVector{battleMap.getCombatantCoordinates(*combatant).getRoot()};
          }
        return {};
      }

    if(!combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return battleMap.getFreeCoordsInHopRange(battleMap.getCombatantCoordinates(_target).get(), distances, combatant->getSize(),
                                                 factory._attackRange, combatant->_instanceId);
      }
    else if(battleMap.getHopDistanceCombatants(*combatant, _target) <= factory._attackRange)
      {
        return CoordVector{battleMap.getCombatantCoordinates(*combatant).getRoot()};
      }

    return {};
  }

  size_t MeleeAttack::hash() const
  {
    size_t h = std::hash<int>{}(static_cast<int>(getAbilityType()));
    h ^= std::hash<int>{}(static_cast<int>(getFlags())) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<int>{}(_target._instanceId) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<std::string>{}(_factory._name) + 0x9e3779b9 + (h << 6) + (h >> 2);
    return h;
  }

  bool MeleeAttack::equals(const Actoid &other) const
  {
    if(auto *meleeAttack = dynamic_cast<const MeleeAttack *>(&other))
      {
        return getAbilityType() == other.getAbilityType() && getFlags() == other.getFlags() && _target._instanceId == meleeAttack->_target._instanceId
               && _factory._name == meleeAttack->_factory._name;
      }
    return false;
  }
}